python/source/flight.rst

```diff
@@ -0,0 +1,228 @@
+            for dataset in self._repo.iterdir():
+                yield self._make_flight_info(dataset.name)
+
+        def get_flight_info(self, context, descriptor):
+            return self._make_flight_info(descriptor.path[0].decode('utf-8'))
+
+        def do_put(self, context, descriptor, reader, writer):
+            dataset = descriptor.path[0].decode('utf-8')
+            dataset_path = self._repo / dataset
+            data_table = reader.read_all()
+            pa.parquet.write_table(data_table, dataset_path)
+
+        def do_get(self, context, ticket):
+            dataset = ticket.ticket.decode('utf-8')
+            dataset_path = self._repo / dataset
+            return pa.flight.RecordBatchStream(pa.parquet.read_table(dataset_path))
+
+        def list_actions(self, context):
+            return [
+                ("drop_dataset", "Delete a dataset."),
+            ]
+
+        def do_action(self, context, action):
+            if action.type == "drop_dataset":
+                self.do_drop_dataset(action.body.to_pybytes().decode('utf-8'))
+            else:
+                raise NotImplementedError
+
+        def do_drop_dataset(self, dataset):
+            dataset_path = self._repo / dataset
+            dataset_path.unlink()
+
+The example server exposes :meth:`pyarrow.flight.FlightServerBase.list_flights`
+which is the method in charge of returning the list of data streams available
+for fetching.
+
+Likewise, :meth:`pyarrow.flight.FlightServerBase.get_flight_info` provides
+the information regarding a single specific data stream.
+
+Then we expose :meth:`pyarrow.flight.FlightServerBase.do_get` which is in charge
+of actually fetching the exposed data streams and sending them to the client.
+
+Allowing to list and dowload data streams would be pretty useless if we didn't
+expose a way to create them, this is the responsability of
+:meth:`pyarrow.flight.FlightServerBase.do_put` which is in charge of receiving
+new data from the client and dealing with it (in this case saving it
+into a parquet file)
+
+This are the most common Arrow Flight requests, if we need to add more
+functionalities, we can do so using custom actions.
+
+In the previous example a `drop_dataset` custom action is added. 
+All custom actions are executed through the 
+:meth:`pyarrow.flight.FlightServerBase.do_action` method, thus it's up to
+the server subclass to dispatch them properly. In this case we invoke
+the `do_drop_dataset` method when the `action.type` is the one we expect.
+
+Our server can then be started with 
+:meth:`pyarrow.flight.FlightServerBase.serve`
+
+.. code-block::
+
+    if __name__ == '__main__':
+        server = FlightServer()
+        server._repo.mkdir(exist_ok=True)
+        server.serve()
+
+.. testcode::
+    :hide:
+
+    # Code block to start for real a server in background
+    # and wait for it to be available.
+    # Previous code block is just to show to user how to start it.
```

Comment:
In that case, should we move the `import threading` from the first code block to here?